src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncStateMachine.cs

```diff
@@ -27,8 +28,8 @@ public AsyncStateMachine(VariableSlotAllocator variableAllocatorOpt, TypeCompila
 
             if (asyncMethod.IsIterator)
             {
-                var elementType = asyncMethod.IteratorElementType;
-                //this.ElementType = TypeMap.SubstituteType(elementType).Type; // PROTOTYPE(async-streams): TODO
+                var elementType = TypeMap.SubstituteType(asyncMethod.IteratorElementType).Type;
+                this.ElementType = elementType;
```

Comment:
:memo: This is the central part of this PR (fixes the async-iterator scenario with a generic type parameter).